src/providers/mbed_provider/asym_encryption.rs

```diff
@@ -0,0 +1,101 @@
+// Copyright 2020 Contributors to the Parsec project.
+// SPDX-License-Identifier: Apache-2.0
+use super::{key_management, MbedProvider};
+use crate::authenticators::ApplicationName;
+use crate::key_info_managers::KeyTriple;
+use log::info;
+use parsec_interface::operations::{psa_asymmetric_decrypt, psa_asymmetric_encrypt};
+use parsec_interface::requests::{ProviderID, ResponseStatus, Result};
+use parsec_interface::secrecy::{ExposeSecret, Secret};
+use psa_crypto::operations::asym_encryption;
+use psa_crypto::types::key;
+
+impl MbedProvider {
+    pub(super) fn psa_asymmetric_encrypt_internal(
+        &self,
+        app_name: ApplicationName,
+        op: psa_asymmetric_encrypt::Operation,
+    ) -> Result<psa_asymmetric_encrypt::Result> {
+        info!("Mbed Provider - Asym Encrypt");
+        let key_name = op.key_name.clone();
+
+        let key_triple = KeyTriple::new(app_name, ProviderID::MbedCrypto, key_name);
+        let store_handle = self.key_info_store.read().expect("Key store lock poisoned");
+        let key_id = key_management::get_key_id(&key_triple, &*store_handle)?;
+        let _guard = self
+            .key_handle_mutex
+            .lock()
+            .expect("Grabbing key handle mutex failed");
+        let id = key::Id::from_persistent_key_id(key_id);
+        let key_attributes = key::Attributes::from_key_id(id)?;
+
+        op.validate(key_attributes)?;
+        let salt_buff = match &op.salt {
+            Some(salt) => Some(salt.as_slice()),
+            None => None,
+        };
+        let alg = op.alg;
+        let buffer_size = key_attributes.asymmetric_encrypt_output_size(alg)?;
+        let mut ciphertext = vec![0u8; buffer_size];
+
+        match asym_encryption::encrypt(
+            id,
+            alg,
+            &op.plaintext.expose_secret(),
+            salt_buff,
+            &mut ciphertext,
+        ) {
+            Ok(output_size) => {
+                ciphertext.resize(output_size, 0);
+                Ok(psa_asymmetric_encrypt::Result { ciphertext })
+            }
+            Err(error) => {
+                let error = ResponseStatus::from(error);
+                format_error!("Encrypt status: {}", error);
```

Comment:
The `format_error` macro doesn't need the `{}` part of the "format string", that gets added inside the macro!

Reply:
Ah, that's why the log outputs have been a bit odd...!

Reply:
What about in this case?
```Rust
format_error!(
       "Error {} when opening a persistent Mbed Crypto key.",
       e
);
```

Reply:
Ooops! I didn't notice that one when it went in... I wrote `format_error` to always log in the same format "Brief description of problem; error: {}" - the last bit is only written when the config flag for verbose logging is on. You can probably change that to `("Failed to open persistent Mbed Crypto key", e)`